File: Python3/401 - 600/519 Random Flip Matrix.py
# 68ms, 98.39%; 13.4MB, 13.3%
class Solution:

    def __init__(self, n_rows: int, n_cols: int):
        self.col = n_cols
        self.maxLength = n_rows * n_cols
        self.pts = set()
        

    def flip(self) -> List[int]:
        while True:
            opt = int(random.random() * self.maxLength)
            if opt not in self.pts: break
        self.pts.add(opt)
        return [opt // self.col, opt % self.col]
        

    def reset(self) -> None:
        self.pts.clear()


# Your Solution object will be instantiated and called as such:
# obj = Solution(n_rows, n_cols)
# param_1 = obj.flip()
# obj.reset()

File: 401 - 600/519 Random Flip Matrix.cc
// 20ms, 100.0%; 12.3MB, 94.44%
#include <bits/stdc++.h>
using namespace std;

static int x = []() {
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);
    return 0;
}();

class Solution {
private:
    int col, maxLength;
    unordered_set<int> pts;

public:
    Solution(int n_rows, int n_cols) : col(n_cols), maxLength(n_rows * n_cols) {}

    vector<int> flip() {
        int opt;
        do {
            opt = rand() % maxLength;
        } while (pts.find(opt) != pts.end());
        pts.insert(opt);
        return vector<int>{opt / col, opt % col};
    }

    void reset() {
        pts.clear();
    }
};

/**
 * Your Solution object will be instantiated and called as such:
 * Solution obj = new Solution(n_rows, n_cols);
 * vector<int> param_1 = obj.flip();
 * obj.reset();
 */
